module/documents/actor/bastion.mjs

```diff
@@ -0,0 +1,401 @@
+   */
+  #evaluateTradeOrder(facility, updates) {
+    const { trade } = facility.system;
+    updates["system.trade.pending.operation"] = null;
+    updates["system.trade.pending.creatures"] = [];
+    updates["system.trade.pending.value"] = null;
+    if ( !trade.pending.operation ) return {};
+
+    if ( trade.pending.operation === "buy" ) {
+      // Stocked facility
+      if ( (trade.pending.value === null) && trade.pending.stocked ) updates["system.trade.stock.stocked"] = true;
+
+      // Bought goods
+      else if ( trade.pending.value !== null && !trade.pending.creatures.length ) {
+        updates["system.trade.stock.value"] = Math.min(trade.stock.value + trade.pending.value, trade.stock.max);
+      }
+    } else if ( trade.pending.value !== null ) {
+      // See OrderActivity#_finalizeTrade for creatures TODO
+      // Sold goods
+      let sold = trade.pending.value;
+      if ( !trade.pending.creatures.length ) {
+        updates["system.trade.stock.value"] = Math.max(0, trade.stock.value - trade.pending.value);
+        sold = trade.stock.value - updates["system.trade.stock.value"];
+      }
+
+      return { gold: Math.floor(sold * ((trade.profit / 100) + 1)) };
+    }
+
+    return {};
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Handle clicking action elements in chat cards.
+   * @param {PointerEvent} event     The triggering event.
+   * @param {HTMLElement} target     The action element.
+   * @param {ChatMessage5e} message  The chat message.
+   */
+  #onChatAction(event, target, message) {
+    const { action } = target.dataset;
+    switch ( action ) {
+      case "claim": this.#onClaimGold(message); break;
+      case "viewItem": this.#onViewItem(target); break;
+    }
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Handle claiming gold from a bastion turn summary message.
+   * @param {ChatMessage5e} message  The message
+   * @returns {Promise<ChatMessage5e>}
+   */
+  async #onClaimGold(message) {
+    const results = message.getFlag("dnd5e", "bastion");
+    const { gold } = results;
+    const actor = message.getAssociatedActor();
+    const { gp } = actor?.system?.currency ?? {};
+    if ( !gold?.value || gold.claimed || (gp === undefined) ) return;
+    await actor.update({ "system.currency.gp": gp + gold.value });
+    gold.claimed = true;
+    const content = await this.#renderSummary(actor, results);
+    return message.update({ content, flags: { dnd5e: { bastion: results } } });
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Handle dragging an item created as part of order completion.
+   * @param {DragEvent} event    The initiating drag event.
+   * @param {Item5e} item        The created item.
+   * @param {object} [updates]   Updates to apply to the Item.
+   */
+  #onDragItem(event, item, updates={}) {
+    // TODO: Need some way to mark the item as 'claimed' when it is dropped onto an Actor sheet.
```

Comment:
We could pass the message ID with the data transfer, and then in `_onDropItemCreate` we can relay that back into the options for `createDocuments` and then have a hook in bastions watching for that item's creation. A bit of a roundabout process, but it would ensure that the message is updated properly even if the player dropping doesn't have permissions on it.

Reply:
I suppose there has to be at least *someone* online who has permission to modify the chat message, right? Or am I misunderstanding the proposal?

Reply:
Well, supposing the messages are created by the GM when they hit Advance Bastion Turn, then they will need to be the one to update it. So players would drag the item to their sheet and send out the `createDocument`, and then the GM would listen for `onCreateDocument` with the message option and update the chat message when it sees that.

Reply:
Yep, got it. I was only considering cases like the players logging in between sessions to do book-keeping and attempting to claim items. But that's an issue with various other things like concentration as well.

Not looking like I'll be able to get to it at this rate, but the TODO is there to come back to.